### PR TITLE
Add apigate proxy response

### PR DESCRIPTION
**Pull request by minsikzzang**
* Add Definitions for Api Gateway Custom Authorizer request and response

More detail
* http://docs.aws.amazon.com/apigateway/latest/developerguide/use-custom-authorizer.html

### CODE REVIEW

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	// The real path corresponding to the path parameters injected into the
+	// Resource placeholders.
+	Path string
+
+	// The incoming request HTTP method name.
+	// Valid values include: DELETE, GET, HEAD, OPTIONS, PATCH, POST, and
+	// PUT.
+	HTTPMethod string
+
+	// The incoming reauest HTTP headers.
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: Typo :wink:

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	APIKey string
+
+	// The source IP address of the TCP connection making the request to
+	// Amazon API Gateway.
+	SourceIP string
+}
+
+// RequestContext provides contextual information about an Amazon API Gateway
+// Proxy Custom Authorizer event.
+//
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz delete this empty comment line

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+// Event represents an input to a REQUEST authorizer for an API method with a
+// proxy integration
+type Event struct {
+	// Type property specifies the authorizer type ("TOKEN" or "REQUEST")
+	Type string
+
+	// MethodARN is the ARN of the incoming method request and is populated by
+	// API Gateway in accordance with the custom authorizer configuration
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz add "." at end of line

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	APIID string
+
+	// The identifier Amazon API Gateway assigns to the resource.
+	ResourceID string
+}
+
+// Event represents an input to a REQUEST authorizer for an API method with a
+// proxy integration
+type Event struct {
+	// Type property specifies the authorizer type ("TOKEN" or "REQUEST")
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz add "." at end of line

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+//
+// More detail about IAM policy statement
+// http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html#api-gateway-calling-api-permissions
+type Statement struct {
+	Action   string
+	Effect   string
+	Resource string
+}
+
+// PolicyDocument represents AWS IAM policy document
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: ... represents **an** AWS IAM ....

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+type PolicyDocument struct {
+	Version string
+
+	// Statement represents AWS IAM policy statement
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: ... represents **an** AWS IAM ....

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+	// The identifier Amazon API Gateway assigns to the API.
+	APIID string
+
+	// The identifier Amazon API Gateway assigns to the resource.
+	ResourceID string
+}
+
+// Event represents an input to a REQUEST authorizer for an API method with a
+// proxy integration
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz  add "." at end of line

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+// Response represents an output from an Amazon API Gateway Custom Authorizer
+type Response struct {
+	// The principal user identification associated with the token sent by the client.
+	PrincipalID string `json:"principalId"`
+
+	// PolicyDocument represents AWS IAM policy document
+	PolicyDocument PolicyDocument `json:"policyDocument"`
+
+	// API Gateway passes the context object from a custom authorizer directly to
+	// the backend Lambda function as part of the input event
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz add "." at end of line

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	// Statement represents AWS IAM policy statement
+	Statement []Statement
+}
+
+// Response represents an output from an Amazon API Gateway Custom Authorizer
+type Response struct {
+	// The principal user identification associated with the token sent by the client.
+	PrincipalID string `json:"principalId"`
+
+	// PolicyDocument represents AWS IAM policy document
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz add "." at end of line and change to "...represents **an** AWS..."

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+// PolicyDocument represents AWS IAM policy document
+type PolicyDocument struct {
+	Version string
+
+	// Statement represents AWS IAM policy statement
+	Statement []Statement
+}
+
+// Response represents an output from an Amazon API Gateway Custom Authorizer
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz add "." at end of line

File: service/lambda/runtime/event/apigatewayauthorizerevt/doc.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+/*
+Package apigatewayproxyevt allows you write AWS Lambda functions as the back end
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/doc.go]: Change for correct description. Like:
Package apigatewayproxyevt allows you to write an Amazon API Gateway custom authorizer that provide control access to your API methods.

File: service/lambda/runtime/event/apigatewayproxyevt/definition.go [hunk 12]
```diff
@@ -151,6 +151,24 @@ type Event struct {
 	RequestContext *RequestContext
 }
 
+// Response represents a API Gateway Lambda Proxy response format
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayproxyevt/definition.go]: plz add "." at end of line, and "...represents **an** API..."

File: service/lambda/runtime/event/apigatewayproxyevt/definition.go [hunk 13]
```diff
@@ -151,6 +151,24 @@ type Event struct {
...
+type Response struct {
+	// A flag to indicate if the applicable request payload is Base64
+	// encoded.
+	IsBase64Encoded bool `json:"isBase64Encoded"`
+
+	// The outgoing HTTP status code
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayproxyevt/definition.go]: plz add "." at end of line

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+	// Type property specifies the authorizer type ("TOKEN" or "REQUEST")
+	Type string
+
+	// MethodARN is the ARN of the incoming method request and is populated by
+	// API Gateway in accordance with the custom authorizer configuration
+	MethodARN string
+
+	// AuthorizationToken has called supplied token which is allow, deny,
+	// unauthorized, or any other string value. An empty string value is the
+	// same as unauthorized (only for TOKEN type)
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: plz add "." at end of line

File: service/lambda/runtime/event/apigatewayproxyevt/definition.go [hunk 15]
```diff
@@ -151,6 +151,24 @@ type Event struct {
 	RequestContext *RequestContext
 }
 
+// Response represents a API Gateway Lambda Proxy response format
+type Response struct {
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayproxyevt/definition.go]: To preserve the good consistency of PR, please make this change in an other PR.

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+	// The source IP address of the TCP connection making the request to
+	// Amazon API Gateway.
+	SourceIP string
+}
+
+// RequestContext provides contextual information about an Amazon API Gateway
+// Proxy Custom Authorizer event.
+//
+type RequestContext struct {
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: The parameter "RequestID" is missing.
```json
    "requestContext": {
        "path": "/request",
        "accountId": "123456789012",
        "resourceId": "05c7jb",
        "stage": "test",
        "requestId": "...",
        "identity": {
            "apiKey": "...",
            "sourceIp": "..."
        },
        "resourcePath": "/request",
        "httpMethod": "GET",
        "apiId": "s4x3opwd6i"
    }
```
Please add it with a short description.

cf. [doc](http://docs.aws.amazon.com/apigateway/latest/developerguide/use-custom-authorizer.html)

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+// http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html#api-gateway-calling-api-permissions
+type Statement struct {
+	Action   string
+	Effect   string
+	Resource string
+}
+
+// PolicyDocument represents AWS IAM policy document
+type PolicyDocument struct {
+	Version string
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: Please add description.

File: service/lambda/runtime/event/apigatewayauthorizerevt/definition.go [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+
+	// The contextual information associated with the API call.
+	RequestContext *RequestContext
+}
+
+// Statement represents AWS IAM policy statement
+//
+// More detail about IAM policy statement
+// http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html#api-gateway-calling-api-permissions
+type Statement struct {
```
Comment (lion3ls) [inline service/lambda/runtime/event/apigatewayauthorizerevt/definition.go]: Please add short description for this 3.

**Review by lion3ls — CHANGES_REQUESTED** [review]

**Comment by minsikzzang** [comment]
@lsuss addressed all comments, I will add another one for adding response in `apigatewayproxyevt`, and found one typo in the `apigatewayproxyevt`, so that will be corrected as well.

**Comment by lion3ls** [comment]
One last thing I just noticed now sry.

Please add README.md for apigatewayauthorizerevt  (check other events). And Update the main README.md to add link to your new supported event.

You did a great work btw :+1: you are the very first to help us by adding an event type so it's a special moment for us you know :sunglasses:.

**Review by lion3ls — APPROVED** [review]

**Pull request merged**